components/browser/state/src/main/java/mozilla/components/browser/state/middleware/InProgressMediaMiddleware.kt

```diff
@@ -0,0 +1,44 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package mozilla.components.browser.state.middleware
+
+import mozilla.components.browser.state.action.BrowserAction
+import mozilla.components.browser.state.action.EngineAction
+import mozilla.components.browser.state.action.MediaSessionAction
+import mozilla.components.browser.state.action.UpdateInProgressMediaTab
+import mozilla.components.browser.state.state.BrowserState
+import mozilla.components.concept.engine.mediasession.MediaSession
+import mozilla.components.lib.state.Middleware
+import mozilla.components.lib.state.MiddlewareContext
+
+/**
+ * [Middleware] that updates [BrowserState.inProgressMediaTabId] everytime the user start playing media
+ * or navigates to another page in the last tab that was playing media.
+ */
+class InProgressMediaMiddleware : Middleware<BrowserState, BrowserAction> {
+    @Suppress("ComplexCondition")
+    override fun invoke(
+        context: MiddlewareContext<BrowserState, BrowserAction>,
+        next: (BrowserAction) -> Unit,
+        action: BrowserAction
+    ) {
+        next(action)
+
+        // Only interested if playing media started in a normal tab
+        if (action is MediaSessionAction.UpdateMediaPlaybackStateAction &&
+            action.playbackState == MediaSession.PlaybackState.PLAYING &&
+            context.state.tabs.firstOrNull {
+                !it.content.private && it.id == action.tabId
+            } != null) {
```

Comment:
Hm, filtering private tabs here is something that doesn't seem obvious. If some consumer is not interested in private tabs then that should maybe be done at a higher layer later?

Reply:
If we don't do this here we'd have the same issue as with the `selectedTabId`.
I'd say to avoid the filtering we could have a `Pair` or two separated parameters for the two tabs types.